### ISSUE TITLE
🐛 takePhoto stuck on step 2/3

### ISSUE THREAD

**Issue by dmandrade** [cuvent/react-native-vision-camera#2372]
### What's happening?

Sometimes the photo taking action gets stuck at step 2/3 and doesn't finish.

### Reproduceable Code

```tsx
const [cameraPosition, setCameraPosition] = useState<'front' | 'back'>('back')
const device = useCameraDevice(cameraPosition,{
    physicalDevices: ['wide-angle-camera']
})

<Camera
    ref={cameraRef}
    device={device}
    format={format}
    hdr={enableHdr}
    isActive={isActive}
    onInitialized={onInitialized}
    onError={onError}
    enableZoomGesture={false}
    orientation="portrait"
    photo={true}
/>
```

### Relevant log output

```shell
2024-01-10T16:12:45.808Z	CameraView	Updating CameraSession...
2024-01-10T16:12:45.809Z	CameraSession	configure { ... }: Waiting for lock...
2024-01-10T16:12:45.809Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=true, outputsChanged=true, sidePropsChanged=true, isActiveChanged=true)
2024-01-10T16:12:45.809Z	CameraSession	isActive is false, skipping CameraDevice configuration.
2024-01-10T16:12:45.809Z	CameraSession	isActive is false, skipping CameraCaptureSession configuration.
2024-01-10T16:12:45.809Z	CameraSession	Successfully updated CameraSession Configuration! isActive: false
2024-01-10T16:12:45.865Z	CameraView	Updating CameraSession...
2024-01-10T16:12:45.866Z	CameraSession	configure { ... }: Waiting for lock...
2024-01-10T16:12:45.866Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=true, outputsChanged=true, sidePropsChanged=true, isActiveChanged=true)
2024-01-10T16:12:45.866Z	CameraSession	Need to rebuild CameraDevice and CameraCaptureSession...
2024-01-10T16:12:45.866Z	CameraSession	Configuring Camera #0...
2024-01-10T16:12:45.870Z	CameraView	Updating CameraSession...
2024-01-10T16:12:45.872Z	CameraDeviceClient	CameraDeviceClient 0: Opened
2024-01-10T16:12:45.874Z	CameraDevices	Camera #0 is now unavailable.
2024-01-10T16:12:45.877Z	CameraView	Updating CameraSession...
2024-01-10T16:12:45.884Z	CameraSession	configure { ... }: Waiting for lock...
2024-01-10T16:12:45.889Z	CameraSession	Successfully configured Camera #0!
2024-01-10T16:12:45.890Z	CameraSession	Destroying previous outputs...
2024-01-10T16:12:45.891Z	CameraSession	Creating outputs for Camera #0...
2024-01-10T16:12:45.892Z	CameraSession	Adding 640x480 Photo Output in JPEG...
2024-01-10T16:12:45.895Z	CameraDevice-JV-0	waitUntilIdle: E. id = 0
2024-01-10T16:12:45.895Z	CameraSession	PreviewView Surface created! Surface(name=null)/@0x88d76fc
2024-01-10T16:12:45.895Z	CameraSession	Setting Preview Output...
2024-01-10T16:12:45.896Z	CameraSession	PreviewView Surface updated! Surface(name=null)/@0x88d76fc 1298 x 2307
2024-01-10T16:12:45.896Z	CameraDevice-JV-0	waitUntilIdle: X
2024-01-10T16:12:45.904Z	CameraView	Updating CameraSession...
2024-01-10T16:12:45.993Z	CameraSession	configure { ... }: Waiting for lock...
2024-01-10T16:12:45.993Z	CameraSession	Successfully configured Session with 1 outputs for Camera #0!
2024-01-10T16:12:45.994Z	CameraSession	CameraSession has no repeating outputs (Preview, Video, CodeScanner), skipping CaptureRequest...
2024-01-10T16:12:45.994Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:45.994Z	CameraView	invokeOnInitialized()
2024-01-10T16:12:45.994Z	CameraView	A new configure { ... } call arrived, aborting this one...
2024-01-10T16:12:45.994Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=false, outputsChanged=false, sidePropsChanged=false, isActiveChanged=false)
2024-01-10T16:12:45.994Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:45.994Z	CameraView	A new configure { ... } call arrived, aborting this one...
2024-01-10T16:12:45.994Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=false, outputsChanged=false, sidePropsChanged=false, isActiveChanged=false)
2024-01-10T16:12:45.994Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:45.994Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=false, outputsChanged=true, sidePropsChanged=true, isActiveChanged=false)
2024-01-10T16:12:45.994Z	CameraSession	Destroying previous outputs...
2024-01-10T16:12:45.998Z	CameraSession	Creating outputs for Camera #0...
2024-01-10T16:12:45.998Z	CameraSession	Adding 640x480 Photo Output in JPEG...
2024-01-10T16:12:45.999Z	CameraSession	Adding 1920x1080 Preview Output...
2024-01-10T16:12:46.001Z	CameraDevice-JV-0	waitUntilIdle: E. id = 0
2024-01-10T16:12:46.001Z	CameraDevice-JV-0	waitUntilIdle: X
2024-01-10T16:12:46.022Z	CameraView	Updating CameraSession...
2024-01-10T16:12:46.023Z	CameraSession	PreviewView Surface updated! Surface(name=null)/@0x88d76fc 1920 x 1080
2024-01-10T16:12:46.166Z	CameraSession	configure { ... }: Waiting for lock...
2024-01-10T16:12:46.166Z	CameraSession	Camera Session android.hardware.camera2.impl.CameraCaptureSessionImpl@cc8597e has been closed.
2024-01-10T16:12:46.166Z	CameraSession	Successfully configured Session with 2 outputs for Camera #0!
2024-01-10T16:12:46.169Z	CameraView	invokeOnStarted()
2024-01-10T16:12:46.169Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:46.169Z	CameraView	A new configure { ... } call arrived, aborting this one...
2024-01-10T16:12:46.169Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=false, outputsChanged=false, sidePropsChanged=false, isActiveChanged=false)
2024-01-10T16:12:46.169Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:46.170Z	CameraSession	configure { ... }: Updating CameraSession Configuration... Difference(deviceChanged=false, outputsChanged=false, sidePropsChanged=false, isActiveChanged=false)
2024-01-10T16:12:46.170Z	CameraSession	Successfully updated CameraSession Configuration! isActive: true
2024-01-10T16:12:47.370Z	CameraView	Finding view 7629...
2024-01-10T16:12:47.370Z	CameraView	Found view 7629!
2024-01-10T16:12:47.370Z	CameraView.takePhoto	Taking photo... Options: {qualityPrioritization=speed, photoCodec=jpeg, enableShutterSound=false, flash=off, quality=40.0}
2024-01-10T16:12:47.370Z	CameraSession	Photo capture 0/3 - preparing capture request (640x480)...
2024-01-10T16:12:47.373Z	CameraSession	Photo capture 1/3 - starting capture...
2024-01-10T16:12:48.147Z	CameraSession	Photo Captured!
2024-01-10T16:12:48.147Z	CameraSession	Photo capture 2/3 complete - received metadata with timestamp 41170456771000
2024-01-10T16:12:48.147Z	CameraSession	Photo captured! 640 x 480
2024-01-10T16:15:00.245Z	CameraView	Updating CameraSession...
...
```

### Camera Device

```json
{
  "sensorOrientation": "landscape-right",
  "hardwareLevel": "full",
  "maxZoom": 10,
  "minZoom": 1,
  "maxExposure": 24,
  "supportsLowLightBoost": false,
  "neutralZoom": 1,
  "physicalDevices": [
    "wide-angle-camera"
  ],
  "supportsFocus": true,
  "supportsRawCapture": true,
  "isMultiCam": false,
  "minExposure": -24,
  "name": "BACK (0)",
  "hasFlash": true,
  "hasTorch": true,
  "position": "back",
  "id": "0"
}
```

### Device

Pocophone M5s (Android 13), Samsung Galaxy A03 (Android 12), Galaxy S22 (Android 13)

### VisionCamera Version

3.7.0

### Can you reproduce this issue in the VisionCamera Example app?

Yes, I can reproduce the same issue in the Example app here

### Additional information

- [X] I am using Expo
- [ ] I have enabled Frame Processors (react-native-worklets-core)
- [X] I have read the [Troubleshooting Guide](https://react-native-vision-camera.com/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

**Comment by mrousavy** [comment]
What logs are coming in after that? You didn't include the logs after.

Maybe you are destroying the Camera session before photo capture has finished? Are you navigating away by any chance?

**Comment by dmandrade** [comment]
I didn't navigate away, the camera remained open and with preview working.

I took the photo and it simply stopped at step 2 of the capture, after some minutes it just generated the logs related to "updating camerasession". I can do other tests and update the issue with complete log output.

The error happens more frequently on weak devices, but I had to repeat the test several times to occur.
It also happens more frequently when goes background and returns before taking a photo

**Comment by stevengoldberg** [comment]
I've observed the same behavior for awhile on my iPhone 14 Pro. I have no idea exactly what conditions cause it, but I have also noticed that it tends to happen when trying to take a photo after the app returns from the background.

I call `const photo = await camera.current.takePhoto(photoOptions)` and the promise never gets resolved, which results in an infinite spinner in my app. The only workaround I've found is to wrap the `takePhoto` call in a function that rejects after a timeout. The hangup still happens intermittently, but when it throws after the timeout it generally works as expected on the next attempt.

**Comment by dolakzdenek** [comment]
For me the issue was that start taking a photo changed `isActive` (through state value) to `false`. It led to the process freeze. When I removed this state changed, it's alright. So check the `isActive` state.

**Comment by stevengoldberg** [comment]
> For me the issue was that start taking a photo changed `isActive` (through state value) to `false`. It led to the process freeze. When I removed this state changed, it's alright. So check the `isActive` state.

FWIW I don’t think we’re talking about the same issue. The issue I’m seeing results in the photo being taken correctly the majority of the time, but intermittently timing out.

**Comment by mrousavy** [comment]
I could do `abortCaptures()` before closing a session to make sure we get a cancelation notice for those photos. I think that's the reason you don't get the callbacks, it's probably because you disable the camera when taking a photo

**Comment by mrousavy** [comment]
Working on a fix for this issue here: https://github.com/mrousavy/react-native-vision-camera/pull/2494 👀 

Still in development, so don't test yet. Long way to go, this is a really complicated issue (as you can see in the changes/Files of that PR)...

**Comment by stevengoldberg** [comment]
Thanks @mrousavy for all your hard work! Is the WIP fix only on the android side? I have this issue on iOS — is that unrelated?

**Comment by mrousavy** [comment]
@stevengoldberg iOS doesn't have this exact issue, there is no such thing as "step 2/3" on iOS. 

So you'd need to create a new issue with logs from iOS for me to investigate it, but for now I'm focused on Android.

**Comment by mrousavy** [comment]
Just released [VisionCamera 3.9.0-beta.1](https://github.com/mrousavy/react-native-vision-camera/releases/tag/v3.9.0-beta.1), which includes the [big `PersistentCaptureSession` PR](https://github.com/mrousavy/react-native-vision-camera/pull/2494) that aims to fix all the blackscreen issues.... This should also now abort captures when switching session, so `takePhoto` will likely not be stuck anymore but will be "safely" aborted :eyes::tada: 

You can try that new beta out yourself and let me know if you still encounter blackscreen issues. I'll continue to test on more devices in the meantime, organize the code a bit better, and then I'll tackle the preview stretching, focus, and flash issues!

If that fixes the blackscreen issues for you, please consider [sponsoring me on GitHub](https://github.com/sponsors/mrousavy) to say thanks! ❤️
